src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs

```diff
@@ -0,0 +1,547 @@
+
+                        Logger.MatchedRoute(entry.RoutePattern.RawText);
+
+                        if (!MatchConstraints(context.HttpContext, values, entry.Constraints))
+                        {
+                            continue;
+                        }
+
+                        await SelectEndpointAsync(context, (Endpoint[])entry.Tag);
+                        if (context.ShortCircuit != null)
+                        {
+                            Logger.RequestShortCircuited(context);
+                            return;
+                        }
+
+                        if (context.Endpoint != null)
+                        {
+                            if (context.Endpoint is IRoutePatternEndpoint templateEndpoint)
+                            {
+                                foreach (var kvp in templateEndpoint.Values)
+                                {
+                                    if (!context.Values.ContainsKey(kvp.Key))
+                                    {
+                                        context.Values[kvp.Key] = kvp.Value;
+                                    }
+                                }
+                            }
+
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool MatchConstraints(HttpContext httpContext, DispatcherValueCollection values, IDictionary<string, IDispatcherValueConstraint> constraints)
+        {
+            if (constraints != null)
+            {
+                foreach (var kvp in constraints)
+                {
+                    var constraint = kvp.Value;
+                    var constraintContext = new DispatcherValueConstraintContext(httpContext, values, ConstraintPurpose.IncomingRequest)
+                    {
+                        Key = kvp.Key
+                    };
+
+                    if (!constraint.Match(constraintContext))
+                    {
+                        values.TryGetValue(kvp.Key, out var value);
+
+                        Logger.RouteValueDoesNotMatchConstraint(value, kvp.Key, kvp.Value);
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        internal Cache CreateCache()
+        {
+            var endpoints = GetEndpoints();
+
+            var groups = new Dictionary<Key, List<Endpoint>>();
+
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var endpoint = endpoints[i];
+
+                var templateEndpoint = endpoint as IRoutePatternEndpoint;
+                if (templateEndpoint == null)
+                {
+                    continue;
+                }
+
+                if (!groups.TryGetValue(new Key(0, templateEndpoint.Pattern), out var group))
+                {
+                    group = new List<Endpoint>();
+                    groups.Add(new Key(0, templateEndpoint.Pattern), group);
```

Comment:
@rynowak This forces the order to 0 so for now none of the tests for `TreeMatcher` check that order is respected.

Reply:
hmm. Is the problem here that an endpoint has no way to convey an 'order'?

Reply:
How about adding a metadata that specifies the order?

Reply:
Will it require an abstraction like `IEndpointOrderMetadata` (or similar) so `GetMetadata` can retrieve it?